File: implementations/countingsort.go
package main

import (
	"fmt"
	"time"
)

func countingSort(arr []int) []int {
	var max int
	var min int = 1 << 32
	for _, e := range arr {
		if e > max {
			max = e
		}
		if e < min {
			min = e
		}
	}
	buckets := make([]int,max-min+1)
	for _, e := range arr {
		buckets[e-min]++
	}

	var out []int
	for i, e := range buckets {
		if e != 0 {
			for j := 0; j < e; j++ {
				out = append(out, i+min)
			}
		}
	}
	return out
}

func main() {
	randarr := []int{9,3,6,2,8,7,3,4}
	start := time.Now()
	countingSort(randarr)
	fmt.Println(time.Since(start))
}

File: python_impls/countingsort.py
import time

def countingsort(arr):
    mi, ma = min(arr), max(arr)
    buckets = [0] * (ma - mi + 1)
    for i in arr:
        buckets[i - mi] += 1
    # use only 1 for loop to return the sorted array
    out = []
    for i in range(len(buckets)):
        out += [i + mi] * buckets[i]
    return out
    
    # return [i + mi for i in range(len(buckets)) for _ in range(buckets[i])]

print(countingsort([5,2,7,9,4,2,6,1,6,9]))